### DISCUSSION TITLE
Quick question about dynamic data transfering

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#1084] [Q&A] [answered]
A located a place in code where dynamic data is transfered:
```c++
earlyTransferTask->transferDynamicData();
```
And inside that function i didn't found any links to size of dynamic array then it was Dirty(). So:
 - Can i change size of array without needing to call compile() again?
 - If yes, can i take some hint where too look at?

**Comment by robertosfield** [comment, marked as answer]
In Vulkan you can't resize GPU resources, so the VSG by design works with fixed size vsg::Data - which is why you won't find and resize methods in vsg::Array, Array2D and Array3D.

If you want to alter the size of resources you'll need to either size the data you create to big enough for all expected usage, or allocate new data and associated buffers and replace the original ones with the new ones.

You don't say anything about what you are doing and what you want to resize so we can't provide any guidance on what might be appropriate in your case,  If you provide more details we should be able to point you in right direction.

**Reply by Anshi701** [reply]
I just saw in copy contructor of array Dirty() with size changed:
```c++
Array(const Array& rhs) :
    Data(rhs.properties, sizeof(value_type)),
    _data(nullptr),
    _size(rhs._size)
{
    if (_size != 0)
    {
        _data = _allocate(_size);
        auto dest_v = _data;
        for (auto& v : rhs) *(dest_v++) = v;
    }
    dirty();
}
```
And thought size changing of array is handled somehow.

**Reply by robertosfield** [reply]
When an vsg::Array object is allocated it also has to allocate the memory that is used to store the data itself, there isn't any resizing going on.

Please, just tell us what you are trying to achieve and we can point you in the right direction, or is it simply that you don't have anything you are trying to do, just trying to figure out how the VSG works?

**Reply by Anshi701** [reply]
Yas, i am tring to understand how memory is handled here. 
RAM memory is allocated, but GPU memory isn't. So it's needs to be compiled before using?
Sorry for bothering you and big thanks for replying!

**Reply by robertosfield** [reply]
The vsg::Data just maps to CPU memory.

To map to GPU memory the VSG uses either vsg::BufferInfo or vsg::ImageInfo. The BufferInfo are used for vertex arrays and descriptor buffers/uniforms.  You can assign the data BufferInfo and leave allocation to the compile traversal, or allocate the space in vk/vsg::Buffer and it's associated vk/vsg::DeviceMemory - this is just what the CompileTraversal does, but you can allocate and set things up any time.

Again, what is the best approach depends upon what you are trying to achieve.

**Reply by Anshi701** [reply]
Thank you! I get it.
